code/game/objects/structures/morgue.dm

```diff
@@ -242,67 +247,65 @@ GLOBAL_LIST_EMPTY(crematoriums)
+///Toggles cremation. The cremate parameter can be used to set a specific value instead of toggling.
+/obj/structure/bodycontainer/crematorium/proc/toggle_cremation(set_cremate, mob/user)
+	if(cremating && set_cremate != TRUE)
+		cremating = FALSE
+		locked = FALSE
+		STOP_PROCESSING(SSobj, src)
+		update_icon()
 
+		audible_message("<span class='hear'>You hear a roar as [src] activates.</span>")
+		return
+	if(!cremating && set_cremate != FALSE)
+		last_user = user
+		cremating = TRUE
+		first_cremation_tick = TRUE
 		locked = TRUE
+		START_PROCESSING(SSobj, src)
 		update_icon()
 
-		for(var/mob/living/M in conts)
-			if (M.stat != DEAD)
-				M.emote("scream")
-			if(user)
-				log_combat(user, M, "cremated")
-			else
-				M.log_message("was cremated", LOG_ATTACK)
-
-			M.death(1)
-			if(M) //some animals get automatically deleted on death.
-				M.ghostize()
-				qdel(M)
-
-		for(var/obj/O in conts) //conts defined above, ignores crematorium and tray
-			qdel(O)
-
-		if(!locate(/obj/effect/decal/cleanable/ash) in get_step(src, dir))//prevent pile-up
-			new/obj/effect/decal/cleanable/ash/crematorium(src)
-
-		sleep(30)
-
-		if(!QDELETED(src))
-			locked = FALSE
-			update_icon()
-			playsound(src.loc, 'sound/machines/ding.ogg', 50, TRUE) //you horrible people
-
-/obj/structure/bodycontainer/crematorium/creamatorium
-	name = "creamatorium"
-	desc = "A human incinerator. Works well during ice cream socials."
+		audible_message("<span class='hear'>The roar of [src]'s flames gradually dies down.</span>")
+		return
 
-/obj/structure/bodycontainer/crematorium/creamatorium/cremate(mob/user)
-	var/list/icecreams = new()
-	for(var/i_scream in GetAllContents(/mob/living))
-		var/obj/item/reagent_containers/food/snacks/icecream/IC = new()
-		IC.set_cone_type("waffle")
-		IC.add_mob_flavor(i_scream)
-		icecreams += IC
-	. = ..()
-	for(var/obj/IC in icecreams)
-		IC.forceMove(src)
+///Ignites and causes heavy burn damage to mobs and objects inside. Non-immune mobs will quickly be incinerated.
+/obj/structure/bodycontainer/crematorium/proc/cremate()
+	if(!cremating)
+		return
+	// Make sure we don't burn the actual morgue and its tray
+	var/list/conts = GetAllContents() - src - connected
+	for(var/mob/living/L in conts)
+		if(first_cremation_tick)
+			if(L.stat != DEAD)
+				L.emote("scream")
+			if(last_user)
+				log_combat(last_user, L, "cremated")
+			else
+				L.log_message("was cremated", LOG_ATTACK)
+			first_cremation_tick = FALSE
+		var/heat_resistant = HAS_TRAIT(L, TRAIT_RESISTHEAT)
+		L.apply_damage(heat_resistant ? 90 : 25, BURN, spread_damage = TRUE)
+		L.adjust_fire_stacks(10)
+		L.IgniteMob()
+
+		//Artificially speed up the cremation process
+		if(iscarbon(L))
+			var/mob/living/carbon/C = L
+			C.adjust_cremation(25)
+
+	for(var/obj/O in conts) //conts defined above, ignores crematorium and tray
+		if(!(O.resistance_flags & FIRE_PROOF))
+			O.fire_act()

```

Comment:
This proc actually takes a temperature var which is very rarely used, but should still be filled out.